Tech/api package

ndelangen opened the pull request:
Issue: Accessing state and API's is kinda bothersome right now.

This extracts the ManagerProvider & ManagerConsumer into a package called `@storybook/api`.
The consumer can now be used in other places.

Also I migrated most of it to TypeScript

ndelangen committed:
REFACTOR context into it's own package && REFACTOR into typescript

ndelangen committed:
FIX tests & linting

ndelangen committed:
CLEANUP

ndelangen commented on lib/api/scripts/generateVersion.js:
```diff
@@ -0,0 +1,9 @@
+const fs = require('fs');
```
I wasn't able to load `package.json` in typescript, since it's outside the `./src` directory.

Generating a `version.ts` upon build felt like an OK solution.

If there's a way to do this in typescript more easily, let me know.

apastuhov replied:
I've tried to remove rootDir and import package.json. Tests passed, but it broke build somehow. Let it be with version generator..

ndelangen commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,211 @@
+  [key: string]: any;
+}
+
+export interface API {
```
Needs work later on, 

Can do that in a separate PR, as it's likely not super duper easy to get right, but obviously immensely useful!

ndelangen commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,211 @@
+  [key: string]: any;
+}
+
+export interface State extends RouterData {
```
Possibly we can combine the substates of all modules, Would love some TypeScript guru help

(I thnk we should do it in a follow-up PR)

ndelangen committed:
SYNC prettier setting for line-length && FIX linting

ndelangen committed:
Merge branch 'next' into tech/api-package

ndelangen committed:
CHANGE the usages of <Consumer> && FIX static-build issue with initialState

ndelangen committed:
IMPROVE treeview search && FIX an issue for build storybook search not working

ndelangen committed:
FIX tests && FIX snapshots && IMPROVE official-example config.js to have filenames

ndelangen committed:
FIX parameters story && FIX HMR in story format

tmeasday commented on examples/official-storybook/config.js:
```diff
@@ -82,21 +82,22 @@ function importAll(context) {
+    const componentOptions =
+      component.prototype && component.prototype.isReactComponent ? { component } : component;
     const kindName = componentOptions.title || componentOptions.component.displayName;
+    const m = componentOptions.module;
```
I was hoping to avoid this

tmeasday replied:
I had HMR working without it?

ndelangen replied:
reverted

tmeasday commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,267 @@
+import Store from './store';
+import getInitialState from './initial-state';
+
+import initAddons from './modules/addons';
```
I think the modules should remain in `lib/ui`

tmeasday replied:
At least the modules that are directly relevant to the UI, such as the layout state.

ndelangen replied:
possibly, do you have a suggestion on how to hook that up?

I feel like there a pros and cons to both approaches. Having the global state in 1 location is not a bad thing?

tmeasday replied:
I guess it depends what you mean. I would expect the state to all be stored in the one "store" but the code for setting/changing the UI's state would live in `lib/ui`. Similarly, addons set their own state in the same store but their store setting state lives in their codebase.

I guess the question is how to make the APIs that the UI creates available to the addons in a principled/typed way. Would it make sense that if an addon wants to alter the UI's state (e.g. set fullscreen) that it should import something from `@storybook/ui`? I feel like that would be sensible.

tmeasday commented on lib/api/src/initial-state.ts:
```diff
@@ -0,0 +1,55 @@
+  viewMode: undefined,
+  storyId: undefined,
+
+  storiesHash: {},
```
The modules should export these themselves

tmeasday replied:
If that's a problem for TS, then we need to solve that, rather than spreading the module over several files, which defeats the purpose of a module.

tmeasday reviewed (changes requested):
This PR seems to combine a few things that we should perhaps evaluate separately:

 1. A bunch of prettier changes to code -- can we try to keep these to separate PRs @ndelangen, they clutter up PRs and make them hard to review. Also .. why do they keep happening? Is my prettier setup wrong? Shouldn't the precommit hook force code to be consistent?

2. Refactoring core APIs to use typescript. Seems like a sensible thing to do but not at this late stage of the release.

3. Refactoring into lib/api. I'm not convinced we should be doing this at this late stage either. There are some unanswered questions, as I've pointed out above.

apastuhov reviewed (commented):
It is hard to review the whole PR, it is too big.

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	addons/cssresources/src/register.tsx
#	addons/notes/src/Panel.tsx
#	addons/notes/src/giphy.tsx
#	examples/official-storybook/stories/core/parameters.stories.js
#	lib/api/src/init-provider-api.ts
#	lib/api/src/lib/id.ts
#	lib/api/src/modules/addons.ts
#	lib/api/src/modules/layout.ts
#	lib/api/src/modules/stories.ts
#	lib/api/src/modules/url.ts
#	lib/api/src/modules/versions.ts
#	lib/api/src/store.ts
#	lib/api/src/tests/id.test.js
#	lib/api/src/tests/versions.test.js
#	lib/client-api/src/client_api.js
#	lib/client-logger/src/index.ts
#	lib/router/package.json
#	lib/router/src/tests/id.test.js
#	lib/router/src/typings.d.ts
#	lib/router/src/utils.ts
#	lib/theming/src/themes/dark-syntax.ts
#	lib/theming/src/themes/dark.ts
#	lib/theming/src/themes/light-syntax.ts
#	lib/theming/src/themes/light.ts
#	lib/ui/package.json
#	lib/ui/src/components/sidebar/treeview/utils.js
#	lib/ui/src/containers/nav.js
#	lib/ui/src/containers/preview.js
#	lib/ui/src/core/context.js
#	lib/ui/src/core/initial-state.js
#	lib/ui/src/index.js
#	lib/ui/src/libs/id.test.js
#	lib/ui/src/settings/__snapshots__/shortcuts.stories.storyshot
#	lib/ui/src/settings/about_page.js
#	lib/ui/src/settings/shortcuts.stories.js
#	lib/ui/src/settings/shortcuts_page.js
#	yarn.lock

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	lib/router/package.json
#	lib/ui/package.json

ndelangen committed:
UPDATE version.ts

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	lib/client-api/src/client_api.js

ndelangen committed:
FIX shortcuts story

ndelangen committed:
FIX an error of storyDataFromString not coping with empty path

ndelangen committed:
FIX issue with SideBar staying in loading state

ndelangen committed:
FIX snapshots

ndelangen committed:
FIX some tests

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	lib/router/package.json
#	lib/ui/package.json

ndelangen committed:
FIX preact acceptance example

ndelangen committed:
FIX version

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	lib/router/package.json
#	lib/ui/package.json

ndelangen committed:
FIX after merge

ndelangen committed:
FIX tests && FIX bootstrap errors

ndelangen committed:
Merge branch 'next' into tech/api-package

ndelangen committed:
IMPROVE type coverage of api

ndelangen committed:
CLEANUP && FIX linting

ndelangen commented:
Merge https://github.com/storybooks/storybook/pull/5978 first to clean this PR up

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	lib/api/src/lib/shortcut.ts
#	lib/router/src/utils.ts

ndelangen committed:
ADD a hook for getting the storybook global state

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	lib/api/src/modules/stories.ts
#	lib/router/package.json
#	lib/ui/package.json
#	yarn.lock

ndelangen committed:
FIX an issue persisted state is empty

ndelangen committed:
Merge branch 'next' into tech/api-package

ndelangen committed:
REVERT changes to config in official example to support HMR

ndelangen committed:
FIX snaphots

ndelangen committed:
Update preact-kitchen-sink

ndelangen committed:
REMOVE the module from example format stories

ndelangen committed:
CLEANUP

ndelangen committed:
CLEANUP

ndelangen committed:
Merge branch 'next' into tech/api-package

ndelangen committed:
CLEANUP

ndelangen committed:
CLEANUP

ndelangen commented:
I cleanup up the massive snapshot change

ndelangen committed:
FIX the out of memory issue

https://github.com/vuejs/vue-cli/issues/1453

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	lib/router/package.json
#	lib/ui/package.json

ndelangen committed:
FIX version

kroeder commented:
cc https://github.com/storybooks/storybook/issues/5030

ndelangen committed:
Merge branch 'next' into tech/api-package

kroeder commented on addons/cssresources/src/css-resource-panel.tsx:
```diff
@@ -5,7 +5,7 @@ import Eventtypes, { STORY_RENDERED } from '@storybook/core-events';
+interface Props {
   active: boolean;
   api: {
     emit(event: any, data: any): void;
```
is `event: any` or is it `event: () => any` ? 
or even `event: (...args) => any` in case it allows properties you don't know  yet?

ndelangen replied:
Hmm that should actually become `API` since there's a type for it now

kroeder commented on addons/cssresources/src/css-resource-panel.tsx:
```diff
@@ -17,12 +17,12 @@ interface CssResourcePanelProps {
 
-export class CssResourcePanel extends Component<CssResourcePanelProps, CssResourcePanelState> {
-  constructor(props: CssResourcePanelProps) {
+export class CssResourcePanel extends Component<Props, State> {
```
Maybe for step 2 but just for showing what's possible if you want to have it 100% accurate
```typescript
interface Props<TEvent, TData> {
  active: boolean;
  api: {
    emit(event: TEvent, data: TData): void;
  };
}

export class CssResourcePanel<TEvent = (...args) => any, TData = any> extends Component<Props<TEvent, TData> , State> {
}
```

Depending on the use case (which I don't know) devs could now make their CssResourcePanel stuff Type-safe if they want to. Default is `any`

ndelangen replied:
Yeah this gives me some idea on how to get this maybe: `Prop<API>`

kroeder commented on addons/notes/src/Panel.tsx:
```diff
@@ -56,6 +56,8 @@ export const SyntaxHighlighter = (props: any) => {
   return <SyntaxHighlighterBase language={language[1]} bordered copyable {...props} />;
```
is `language[1]` possibly null and should be checked?
```typescript
if(!language[1]) {
  throw new Error("...");
}
return <SyntaxHighlighterBase language={language[1]} bordered copyable {...props} />;
```

kroeder commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,255 @@
+  handleAPI(api: API): void;
+  renderPreview(): void;
+  getElements(type: Types): Collection;
+  [key: string]: any;
```
Can this `any` be solved by a generic and does it make sense?
`interface Provider<T> { [key:string]: T; }`

kroeder commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,255 @@
+
+    const store = new Store({
+      getState: () => this.state,
+      setState: (a: any, b: any) => this.setState(a, b),
```
Can you give more context on `a: any, b: any`? Maybe rename the props?

ndelangen replied:
Maybe I use the `typeof this.setState`?

kroeder commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,255 @@
+    });
+
+    api.on(SET_STORIES, (data: any) => {
+      api.setStories(data.stories);
```
You are accessing `data.stories` even though you use `data: any` 
So you at least know at this point you have `data: { stories: SomeStoryType }`

ndelangen replied:
Good catch this object is well known and I can create a type for it, I may already have it.

kroeder commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,255 @@
+
+interface ConsumerProps<A> {
+  filter: (c: Combo) => A;
+  children: (d: A) => React.ReactElement<any>;
```
Can you give more context to the names `<A>`, `c`, `d`?

ndelangen replied:
sure

kroeder commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,255 @@
+  Object.entries(obj).reduce((acc, [k, v]) => acc.concat([k, v]), [])
+);
+
+const ArrayToObject = memoize(100)((array: any[]) =>
```
Is this what you want to achieve?
```typescript
const myArray = ['foo', 'bar', 'baz', { complex: 1000 }];
const myObject = { ...myArray };
console.log(myObject);
```
![image](https://user-images.githubusercontent.com/5478899/54181772-edb40e80-449f-11e9-9f08-3e7a769414bf.png)

ndelangen replied:
I wrote 2 functions that do each other's opposite in order to memoize functions that take an single object as argument.

kroeder commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,255 @@
+  }, {})
+);
+
+class ManagerConsumer extends Component<ConsumerProps<any>> {
```
Worth using `class ManagerConsumer<TRenderMemory = any, TDataMemory = any>` ?

kroeder commented on lib/api/src/index.tsx:
```diff
@@ -0,0 +1,255 @@
+
+  constructor(props: ConsumerProps<any>) {
+    super(props);
+    this.renderMemory = memoize(10)((...args: any[]) => props.children(ArrayToObject(args)));
```
As mentioned above, have you tried using 
```typescript
this.renderMemory = memoize(10)((...args: any[]) => props.children({ ...args }));
``` 
?

kroeder commented on lib/api/src/modules/channel.ts:
```diff
@@ -0,0 +1,39 @@
+
+type CallBack = (...args: any[]) => any;
+
+export interface SubAPI {
```
Worth adding generic types for all those `any` ?

ndelangen replied:
I'm not sure i understand this one.

benoitdion replied:
not sure if that's what @kroeder is suggesting but it'd be nice to have an option for strongly typed events so the callback can be typed.

ndelangen replied:
Sounds cool, shall we deal with that in a follow up PR?

benoitdion replied:
Definitely outside the scope of this PR

kroeder commented on lib/api/src/modules/shortcuts.ts:
```diff
@@ -23,17 +65,26 @@ export const defaultShortcuts = Object.freeze({
+    addEventListener(): void;
+    removeEventListener(): boolean;
+    dispatchEvent(event: Event): boolean;
+    getAttribute(attr: string): string | null;
```
Is it `null`? Or `undefined`? `null` is treated as a value, `undefined` is just not defined

ndelangen replied:
it DOM-api returns either `null` or `string`

kroeder reviewed (commented):
Did a review. Many comments are redundant 🙂

ndelangen committed:
IMPROVE typing of lib/addon

ndelangen committed:
CLEANUP

ndelangen committed:
connect addons/a11y to the API type

ndelangen committed:
connect addons/cssresurces to the API type

ndelangen committed:
connect addons/actions to Theme & API types

ndelangen committed:
connect addons/background to API type

ndelangen committed:
WIP

ndelangen committed:
CLEANUP

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	addons/a11y/package.json
#	addons/actions/package.json
#	addons/backgrounds/package.json
#	addons/cssresources/package.json
#	addons/notes/package.json
#	lib/addons/package.json
#	lib/router/package.json
#	lib/ui/package.json

ndelangen committed:
FIX provider

ndelangen committed:
FIX broken provider

ndelangen committed:
FIX an issue with background tool && MIGRATE it to use the new Consumer with a non-pure filter

ndelangen committed:
CLEANUP types for addon-background

shilman committed:
Merge pull request #6046 from thekogmo/patch-3

Update index.md

ndelangen committed:
DISABLE teamcity running of unit-tests

These exact tests are already run in CircleCI and teamcity seems to timeout on them.

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	addons/a11y/package.json
#	addons/actions/package.json
#	addons/backgrounds/package.json
#	addons/cssresources/package.json
#	addons/notes/package.json
#	lib/addons/package.json
#	lib/api/src/modules/layout.ts
#	lib/api/src/modules/shortcuts.ts
#	lib/ui/package.json

ndelangen committed:
use the filter in all consumers where applicable

ndelangen committed:
FIX merge conflict

ndelangen committed:
Revert "DISABLE teamcity running of unit-tests"

This reverts commit a7cd8c3d62c610b1825a5fa431251d6b0cffc352.

ndelangen committed:
FIX teamcity speed?

ndelangen committed:
Merge branch 'next' into tech/api-package

shilman committed:
Merge pull request #6084 from raybooysen/patch-3

Minor grammar change

ndelangen committed:
CLEANUP addon-notes && use Consumer & filter

ndelangen committed:
ADD the version module and url module to the api type

ndelangen committed:
CHANGE addon-events to use API instead of channel

ndelangen committed:
FIX missing Version export type

ndelangen committed:
CLEANUP addon-notes

shilman committed:
Merge pull request #6105 from danielsogl/patch-3

docs(configurations): fix markdown for react-typescript

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	lib/ui/src/containers/preview.js
#	lib/ui/src/core/init-provider-api.js
#	lib/ui/src/provider.js
#	yarn.lock

breadadams committed:
Update Vue "manual setup" steps

+ Include `babel-preset-vue` in the required dependency instructions

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	yarn.lock

ndelangen committed:
FIX background transparent on Preview causing the addons to be visisble, when they are not supposed to be

Thanks to glitchperfect's https://github.com/storybooks/storybook/pull/6042

ndelangen committed:
Update tsconfig.json

ndelangen committed:
REMOVE the pure functionality as it's not really usable && ADD overloads for setState && ADD types for eventhandlers

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	addons/backgrounds/src/containers/BackgroundSelector.tsx

ndelangen committed:
FIX issue with backgroundSelector

benoitdion commented:
This will hopefully help restore/simplify some of the ondevice addons. ondevice-backgrounds is mostly broken at the moment because it's still expecting data coming through the websocket when some of the values are in state.

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	addons/a11y/package.json
#	addons/actions/package.json
#	addons/backgrounds/package.json
#	addons/cssresources/package.json
#	addons/cssresources/src/css-resource-panel.tsx
#	addons/notes/package.json
#	app/react-native-server/package.json
#	app/react-native-server/src/client/manager/provider.js
#	examples-native/crna-kitchen-sink/package.json
#	lib/addons/package.json
#	lib/ui/package.json
#	yarn.lock

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	addons/a11y/package.json
#	addons/actions/package.json
#	addons/backgrounds/package.json
#	addons/cssresources/package.json
#	addons/notes/package.json
#	app/react-native-server/package.json
#	lib/addons/package.json
#	lib/api/src/modules/versions.ts
#	lib/ui/package.json
#	lib/ui/src/core/context.js

ndelangen commented:
Seems these tests may be leaking memory:
- `app/react/src/client/preview/index.test.js`
- `app/angular/src/client/preview/angular/decorators.test.ts`

ndelangen committed:
FIX types

ndelangen committed:
REMOVE much of the need for channel in addon-links

ndelangen committed:
ADD once functionality to api && FIX types for provider.channel

ndelangen committed:
ADD script for detecting memoryLeaks in tests

ndelangen committed:
CLEANUP && FIX tests

ndelangen committed:
ADD always mock for util-deprecate && REMOVE loglines in tests for debug && CLEANUP tests log

ndelangen commented:
Would like to investigate why some tests take so long
```
 PASS  lib/ui/src/components/sidebar/treeview/utils.test.js
 PASS  lib/client-api/src/client_api.test.js (5.282s)
 PASS  lib/client-api/src/story_store.test.js
 PASS  lib/core/src/server/presets.test.js
 PASS  lib/cli/lib/add.test.js
 PASS  addons/storysource/src/loader/inject-decorator.test.js (6.218s)
 PASS  addons/knobs/src/components/__tests__/Panel.js (12.486s)
 PASS  lib/ui/src/settings/shortcuts.test.js
 PASS  lib/channels/src/index.test.ts (14.99s)
 PASS  lib/api/src/tests/stories.test.js (15.61s)
 PASS  lib/api/src/tests/versions.test.js (15.903s)
 PASS  app/riot/src/client/preview/render-riot.test.js
 PASS  lib/api/src/tests/url.test.js
 PASS  lib/theming/src/tests/create.test.js
 PASS  lib/api/src/tests/shortcut.test.js
 PASS  app/react/src/server/cra-config.test.js
 PASS  addons/knobs/src/components/__tests__/Options.js
 PASS  app/react/src/client/preview/element_check.test.js
 PASS  lib/components/src/typography/link/link.test.js
 PASS  addons/cssresources/src/css-resource-panel.test.js (21.496s)
 PASS  addons/info/src/components/PropTable.test.js
 PASS  addons/info/src/components/PropTable/index.test.js
 PASS  lib/client-api/src/subscriptions_store.test.js
 PASS  addons/info/src/index.test.js (6.35s)
 PASS  lib/api/src/tests/shortcuts.test.js
 PASS  addons/knobs/src/KnobManager.test.js
 PASS  lib/core/src/client/preview/start.test.js
 PASS  addons/actions/src/preview/__tests__/action.test.js
 PASS  lib/api/src/tests/store.test.js (13.492s)
 PASS  addons/storyshots/storyshots-puppeteer/src/__tests__/url.test.js
 PASS  app/react/src/server/framework-preset-react-docgen.test.js
 PASS  app/angular/src/client/preview/angular/decorators.test.ts
 PASS  lib/core/src/server/utils/merge-webpack-config.test.js
 PASS  lib/router/src/tests/id.test.js
 PASS  addons/storyshots/storyshots-core/src/Stories2SnapsConverter.test.js
 PASS  addons/a11y/src/components/A11YPanel.test.js (18.229s)
 PASS  lib/client-logger/src/index.test.ts
 PASS  addons/knobs/src/components/__tests__/Array.js
 PASS  lib/addons/src/make-decorator.test.ts (12.83s)
 PASS  addons/links/src/preview.test.js
 PASS  addons/info/src/components/PropTable/components/Td.test.js
 PASS  addons/ondevice-notes/src/__tests__/index.js
 PASS  lib/core/src/server/utils/template.test.js
 PASS  addons/knobs/src/components/__tests__/RadioButtons.js
 PASS  addons/links/src/react/components/link.test.js
 PASS  addons/knobs/src/components/__tests__/Select.js
 PASS  lib/core/src/server/utils/interpret-files.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.configFunc.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.async.test.js

 PASS  app/angular/src/server/ts_config.test.js
 PASS  lib/api/src/tests/notifications.test.js
 PASS  lib/cli/lib/detect.test.js

 RUNS  addons/notes/src/Panel.test.js
 PASS  app/angular/src/server/create-fork-ts-checker-plugin.test.js
 PASS  app/react/src/client/preview/index.test.js
 PASS  app/angular/src/server/angular-cli_config.test.js
 PASS  lib/node-logger/src/index.test.ts
 PASS  addons/info/src/components/PropTable/components/Tr.test.js
 PASS  addons/info/src/components/PropTable/components/Th.test.js
 PASS  addons/info/src/components/PropTable/components/Table.test.js
 PASS  addons/info/src/components/PropTable/components/Tbody.test.js
 PASS  addons/info/src/components/PropTable/components/Thead.test.js
 PASS  lib/core-events/src/index.test.ts
 PASS  examples/svelte-kitchen-sink/src/components/Button.test.js
 PASS  lib/client-api/src/pathToid.test.js
 PASS  addons/actions/src/preview/__tests__/configureActions.test.js
 PASS  addons/notes/src/Panel.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.specificConfig.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.enzyme.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.test.js
 PASS  lib/cli/lib/project_types.test.js
 PASS  examples/cra-kitchen-sink/src/storyshots.test.js (7.836s)
 PASS  examples/preact-kitchen-sink/preactshots.test.js
 PASS  examples/html-kitchen-sink/tests/htmlshots.test.js (5.737s)
 PASS  examples/svelte-kitchen-sink/svelteshots.test.js (5.415s)
 PASS  lib/ui/src/__tests__/index.js (6.594s)
 PASS  examples/cra-kitchen-sink/src/App.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.shallowWithOptions.test.js
 PASS  lib/codemod/src/transforms/__tests__/move-buildin-addons.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.snapshotWithOptionsFunction.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.renderWithOptions.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.renderOnly.test.js
 PASS  lib/codemod/src/transforms/__tests__/update-organisation-name.test.js
 PASS  addons/storyshots/storyshots-core/stories/storyshot.shallow.test.js
 PASS  examples/riot-kitchen-sink/riotshots.test.js
 PASS  lib/codemod/src/transforms/__tests__/update-addon-info.test.js
 PASS  examples/vue-kitchen-sink/vueshots.test.js
 PASS  examples/angular-cli/angularshots.test.js (13.303s)
 PASS  examples/official-storybook/tests/storyshots.test.js (22.102s)
```

ndelangen committed:
trying to speed up yarn tests

ndelangen committed:
Merge branch 'next' into tech/api-package

# Conflicts:
#	yarn.lock

ndelangen commented on .circleci/config.yml:
```diff
@@ -264,7 +264,7 @@ jobs:
       - run:
           name: Test
-          command: yarn test --coverage --runInBand --core
+          command: yarn test --coverage --w2 --core
```
This speeds up the processing of coverage a lot. It's limited to 2 CPUs since that seems to work on CIs best.

ndelangen committed:
FIX yarn.lock file for docs

ndelangen commented on addons/info/src/__snapshots__/index.test.js.snap:
```diff
@@ -1,7 +1,7 @@
 
 exports[`addon Info should render <Info /> and external markdown 1`] = `
-<deprecated>
+<Component>
```
Side-effect of muting the deprecation messages when running tests

ndelangen commented on addons/notes/src/Panel.tsx:
```diff
@@ -65,87 +70,61 @@ const defaultOptions = {
-  componentWillUnmount() {
-    const { api } = this.props;
-    api.off(STORY_RENDERED, this.onStoryChange);
+const NotesPanel = ({ active }: Props) => {
```
Addon became state-less 🎉

ndelangen commented on jest.config.js:
```diff
@@ -35,17 +35,23 @@ module.exports = {
   snapshotSerializers: ['jest-emotion', 'enzyme-to-json/serializer'],
   coverageDirectory: 'coverage',
-  testEnvironment: 'jsdom',
+  coverageReporters: ['lcov'],
```
This seems to improve the coverage report massively as well

ndelangen commented on lib/addons/src/make-decorator.test.ts:
```diff
@@ -11,15 +10,16 @@ export const defaultDecorateStory = (getStory: StoryGetter, decorators: Decorato
-    warning,
-  });
-  return deprecatedFn;
+jest.mock('util-deprecate', () => {
```
`deprecate` is already mocked, but this changes the implementation

ndelangen commented on lib/api/src/lib/shortcut.ts:
```diff
@@ -1,7 +1,7 @@
 
 // The shortcut is our JSON-ifiable representation of a shortcut combination
-type Shortcut = string[];
+import { KeyCollection, Event } from '../modules/shortcuts';
```
@Jessica-Koch I'm hoping I did the merges correctly, It seems to all work. I hope this is in line with how you want it.

It's all in typescript now 🎉

Jessica-Koch commented:
Omg
Awesome!!!! 

Sent from my iPhone

> On Mar 18, 2019, at 5:03 PM, Norbert de Langen <notifications@github.com> wrote:
> 
> @ndelangen commented on this pull request.
> 
> In lib/api/src/lib/shortcut.ts:
> 
> > @@ -1,7 +1,7 @@
>  import { navigator } from 'global';
>  
>  // The shortcut is our JSON-ifiable representation of a shortcut combination
> -type Shortcut = string[];
> +import { KeyCollection, Event } from '../modules/shortcuts';
> @Jessica-Koch I'm hoping I did the merges correctly, It seems to all work. I hope this is in line with how you want it.
> 
> It's all in typescript now 🎉
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

Hypnosphi commented on lib/api/src/modules/layout.ts:
```diff
@@ -1,67 +1,135 @@
-const checkDeprecatedThemeOptions = options => {
-  if (Object.keys(deprecatedThemeOptions).find(key => !!options[key])) {
+const checkDeprecatedThemeOptions = (options: Options) => {
+  if (Object.values(deprecatedThemeOptions).find(v => !!v)) {
```
What is this check intended to do? Looks like it just checks whether `deprecatedThemeOptions` has some values, which is always true

ndelangen replied:
it should map deprecated options to themeVars

```
name: "brandTitle";
url: "brandUrl";
```

Hypnosphi replied:
I mean that
```
if (Object.values({name: 'brandTitle', url: 'brandUrl'}).find(v => !!v))
```

evaluates to
```
if (['brandTitle', 'brandUrl'].find(v => !!v))

// same as
if ('brandTitle')

// same as
if (true)
```
